src/java/org/apache/cassandra/config/registry/ConfigurationRegistry.java

```diff
@@ -0,0 +1,371 @@
+                                          e -> new AbstractMap.SimpleEntry<>(e.getKey(), e.getValue().getValue()));
+    }
+
+    /**
+     * Setter for the property with the given name. Can accept {@code null} value.
+     * @param name the name of the property.
+     * @param value the value to set.
+     */
+    public void update(String name, Object value)
+    {
+        ConfigurationUnit<?> unit = registry.get(name);
+
+        if (unit == null)
+            throw new ConfigurationException(String.format("Property with name '%s' is not availabe.", name));
+
+        if (unit.setter == null)
+            throw new ConfigurationException(String.format("Property with name '%s' is not writable.", name));
+
+        unit.lock.lock();
+        try
+        {
+            Object oldValue = unit.getValue();
+            if (listeners.get(name) != null)
+                listeners.get(name).onBeforeChange(name, oldValue, value);
+
+            unit.setValue(value);
+            // This potentially may expose the values that are not safe to see in logs on production.
+            logger.info("Updated property {} from {} to {}", name, oldValue, value);
+        }
+        catch (Exception e)
+        {
+            throw new ConfigurationException(String.format("Error updating property with name '%s', cause: %s", name, e.getMessage()), e);
+        }
+        finally
+        {
+            unit.lock.unlock();
+        }
+    }
+
+    public Class<?> getPropertyType(String key)
+    {
+        if (!registry.containsKey(key))
+            throw new ConfigurationException(String.format("Property with name '%s' is not availabe.", key));
+
+        return registry.get(key).getter.getType();
+    }
+
+    public String getBackwardCompatibleKey(String key)
+    {
+        return BACKWARDS_COMPATABLE_NAMES.get(key);
+    }
+
+    /**
+     * Adds a listener for the property with the given name.
+     * @param name property name to listen to.
+     * @param listener listener to add.
+     * @param <T> type of the property.
+     */
+    public <T> void addPropertyChangeListener(String name, PropertyChangeListener<T> listener, Class<T> listenerType)
+    {
+        ConfigurationUnit<?> unit = registry.get(name);
+        if (unit == null)
+            throw new ConfigurationException(String.format("Property with name '%s' is not availabe for resitering handler.", name));
+
+        if (unit.setterType == null)
+            throw new ConfigurationException(String.format("Property with name '%s' is not writable.", name));
+
+//        Class<T> type = (Class<T>) ((ParameterizedType) listener.getClass().getGenericInterfaces()[0]).getActualTypeArguments()[0];
+        if (!unit.setterType.equals(listenerType))
+            throw new ConfigurationException(String.format("Property with name '%s' expects type '%s', but got '%s'.", name, unit.setterType, listenerType));
+
+        listeners.put(name, new PropertyChangeListenerWrapper<>(listener, listenerType));
```

Comment:
what happens if multiple different places need to listen to the same config key?